src/pyhf/optimize/common.py

```diff
@@ -0,0 +1,129 @@
+
+
+def shim(
+    objective,
+    data,
+    pdf,
+    init_pars,
+    par_bounds,
+    fixed_vals=None,
+    do_grad=False,
+    do_stitch=False,
+):
+    """
+    Prepare Minimization for Optimizer.
+
+    Args:
+        objective (`func`): objective function
+        data (`list`): observed data
+        pdf (`pyhf.pdf.Model`): model
+        init_pars (`list`): initial parameters
+        par_bounds (`list`): parameter boundaries
+        fixed_vals (`list`): fixed parameter values
+
+    .. note::
+
+        ``minimizer_kwargs`` is a dictionary containing
+
+          - ``func`` (`func`): backend-wrapped ``objective`` function
+          - ``x0`` (`list`):  modified initializations for minimizer
+          - ``jac`` (`func`): callable that accepts same parameters as the input ``objective`` but returns the gradient
+          - ``bounds`` (`list`): modified bounds for minimizer
+          - ``fixed_vals`` (`list`): modified fixed values for minimizer
+
+    .. note::
+
+        ``stitch_pars(pars, stitch_with=None)`` is a callable that will
+        stitch the fixed parameters of the minimization back into the unfixed
+        parameters.
+
+    .. note::
+
+        ``do_stitch`` will modify the ``init_pars``, ``par_bounds``, and ``fixed_vals`` by stripping away the entries associated with fixed parameters. The parameters can be stitched back in via ``stitch_pars``.
+
+    Returns:
+        minimizer_kwargs (`dict`): arguments to pass to a minimizer following the ``scipy.optimize.minimize`` API (see notes)
+        stitch_pars (`func`): callable that stitches fixed parameters into the unfixed parameters
+    """
+    tensorlib, _ = get_backend()
+
+    if do_stitch:
+        all_init = default_backend.astensor(init_pars)
+        all_idx = default_backend.astensor(range(pdf.config.npars), dtype='int')
+
+        fixed_vals = fixed_vals or []
+        fixed_values = [x[1] for x in fixed_vals]
+        fixed_idx = [x[0] for x in fixed_vals]
+
+        variable_idx = [x for x in all_idx if x not in fixed_idx]
+        variable_init = default_backend.tolist(all_init[variable_idx])
+        variable_bounds = [par_bounds[i] for i in variable_idx]
+        # stitched out the fixed values, so we don't pass any to the underlying minimizer
+        minimizer_fixed_vals = []
+
+        tv = _TensorViewer([fixed_idx, variable_idx])
+        # NB: this is a closure, tensorlib needs to be accessed at a different point in time
+        def stitch_pars(pars, stitch_with=fixed_values):
+            tb, _ = get_backend()
+            return tv.stitch([tb.astensor(fixed_values, dtype='float'), pars])
+
+    else:
+        tv = None
+        variable_init = init_pars
+        variable_bounds = par_bounds
+        minimizer_fixed_vals = fixed_vals
+        stitch_pars = lambda pars, stitch_with=None: pars
+
+    objective_and_grad = _get_tensor_shim()(
+        objective, tensorlib.astensor(data), pdf, stitch_pars, do_grad=do_grad,
```

Comment:
shouldn't `data` already be a tensor? do we need the cast?

Reply:
`data` didn't need to be a tensor. I can change it to be a tensor though. There were existing tests written that allowed non-tensor data (just a list) e.g. `[10.] + pdf.config.auxdata`.

Reply:
ok

Reply:
#980 is opened for this.